filters/block/matcher.go

```diff
@@ -0,0 +1,198 @@
+package block
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type maxBufferHandling int
+
+const (
+	readBufferSize uint64 = 8192
+)
+
+const (
+	maxBufferBestEffort maxBufferHandling = iota
+	maxBufferAbort
+)
+
+// matcher provides a reader that wraps an input reader, and blocks the request
+// if a pattern was found.
+//
+// It reads enough data until at least a complete match of the
+// pattern is met or the maxBufferSize is reached. When the pattern matches the entire
+// buffered input, the replaced content is returned to the caller when maxBufferSize is
+// reached. This also means that more replacements can happen than if we edited the
+// entire content in one piece, but this is necessary to be able to use the matcher for
+// input with unknown length.
+//
+// When the maxBufferHandling is set to maxBufferAbort, then the streaming is aborted
+// and the rest of the payload is dropped.
+//
+// To limit the number of repeated scans over the buffered data, the size of the
+// additional data read from the input grows exponentially with every iteration that
+// didn't result with any matched data blocked. If there was any matched data
+// the read size is reset to the initial value.
+//
+// When the input returns an error, e.g. EOF, the matcher finishes matching the buffered
+// data, blocks or return it to the caller.
+//
+// When the matcher is closed, it doesn't read anymore from the input or return any
+// buffered data. If the input implements io.Closer, closing the matcher closes the
+// input, too.
+//
+type matcher struct {
+
+	// init:
+	input             io.ReadCloser
+	matchList         []string
```

Comment:
this should be either a map[int][]byte or a []buffer with `type buffer []byte`
see also https://go.dev/play/p/IKAa6eA91YD
I think the type is more memory efficient, but it does not really matter if we do not store millions of strings.